sympy/functions/special/holonomic.py

```diff
@@ -0,0 +1,523 @@
+    and the parent ring which must be an instance of DifferentialOperatorAlgebra.
+
+    An Operator can also be created easily using the operator `Dx`. See examples below.
+
+    Examples
+    ========
+
+    >>> from sympy.functions.special.holonomic import DifferentialOperator, DiffOperatorAlgebra
+    >>> from sympy.polys.domains import ZZ, QQ
+    >>> from sympy import symbols
+    >>> x = symbols('x')
+    >>> R, Dx = DiffOperatorAlgebra(ZZ.old_poly_ring(x),'Dx')
+
+    >>> DifferentialOperator([0,1,x**2], R)
+    (1)Dx + (x**2)Dx**2
+
+    One can use the `Dx` and represent differential operators more conveniently this way also.
+    >>> (x*Dx*x + 1 - Dx**2)**2
+    (2*x**2 + 2*x + 1) + (4*x**3 + 2*x**2 - 4)Dx + (x**4 - 6*x - 2)Dx**2 + (-2*x**2)Dx**3 + (1)Dx**4
+
+    """
+
+    _op_priority = 30
+
+    def __init__(self, list_of_poly, parent, generator=None, variable=None):
+        # the parent ring for this operator
+        # must be an DifferentialOperatorAlgebra object
+        self.parent = parent
+        # the variable (or generator) of the base polynomial ring
+        if variable is None:
+            self.variable = symbols('x')
+        else:
+            self.variable = variable
+        # sequence of polynomials in x for each power of Dx
+        # represents the operator
+        # convert the expressions into ring elements using from_sympy
+        if isinstance(list_of_poly, list):
+            for i, j in enumerate(list_of_poly):
+                if isinstance(j, int):
+                    list_of_poly[i] = (
+                        (self.parent).base).from_sympy(S(j))
+                elif not isinstance(j, self.parent.base.dtype):
+                    list_of_poly[i] = (
+                        (self.parent).base).from_sympy(j)
+
+            self.listofpoly = list_of_poly
+        self.order = len(self.listofpoly) - 1
+
+    def args(self):
+
+        return self.gen_symbol
+
+    def __mul__(self, other):
+        """
+        Multiplies two DifferentialOperator and returns another
+        DifferentialOperator isinstance using the commutation rule
+        Dx*a = a*Dx + a'
+        """
+
+        if isinstance(other, DifferentialOperator):
+            # if other is `Dx`
+            if other.listofpoly == other.parent.derivative_operator.listofpoly:
+                sol = []
+                sol.append(other.parent.base.zero)
+                for i in self.listofpoly:
+                    sol.append(i)
+
+                return DifferentialOperator(sol, self.parent)
+
+        gen = self.parent.derivative_operator
+        listofpoly = self.listofpoly
+        for i, j in enumerate(listofpoly):
+            if isinstance(j, self.parent.base.dtype):
+                listofpoly[i] = self.parent.base.to_sympy(j)
+        sol = (listofpoly[0] * other)
+        # using the commutation rule gen*b = b*gen + b'
+
+        def _diff_n_times(b):
+            expr = b * gen + b.diff()
```

Comment:
Okay, I found a method where we don't have convert the list to expressions but it will still contain the special case of `other == Dx`. Is something like this would be efficient.